Added FarmersHoe item with reseeding

Rene98 opened the pull request.

Rene98 committed:
Added FarmersHoe item with reseeding

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/SupServItemSetup.java:
```diff
@@ -76,6 +78,14 @@ public static void setup(@Nonnull SupServPlugin plugin) {
+                        SlimefunItems.EARTH_RUNE, SlimefunItems.WATER_RUNE, SlimefunItems.EARTH_RUNE,
+                        new ItemStack(Material.BONE_BLOCK), new ItemStack(Material.IRON_HOE), new ItemStack(Material.BONE_BLOCK),
+                        SlimefunItems.EARTH_RUNE, SlimefunItems.WATER_RUNE, SlimefunItems.EARTH_RUNE,
+                }, new SlimefunItemStack(SupServItems.FARMERS_HOE, 1)
```
You don't need to specify the recipe output unless the stack size is greater than 1 or the item is not the same as specified in the second parameter.

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/SupServItems.java:
```diff
@@ -72,4 +72,10 @@ private SupServItems() {}
 
+    public static final SlimefunItemStack FARMERS_HOE = new SlimefunItemStack("FARMERS_HOE",
+            Material.IRON_HOE,
+            "&3Farmers&8Hoe",
```
Usually I like to keep my names single colored, but this can be decided later

NCBPFluffyBear replied:
Can we make this `&3Farmers Hoe`

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/SupServItems.java:
```diff
@@ -72,4 +72,10 @@ private SupServItems() {}
+            "&3Farmers&8Hoe",
+            "",
+            "&7Makes sure harvested plots get reseeded, at the cost of durability",
+            "&9Suggested by Caveman");
```
Add a new empty line before the suggestion, and fix formatting by moving the closing parentheses + semicolon to a new line

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,61 @@
+
+    }
+
+    List<Material> farmables = Arrays.asList(Material.DIRT, Material.GRASS_BLOCK, Material.COARSE_DIRT);
```
Should be a class constant

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class FarmersHoe  extends SimpleSlimefunItem<ItemUseHandler> {
```
A know issue with ItemUseHandler items is that it causes the block its used on to become the item, implement NotPlaceable to fix this

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,61 @@
+            Player p = e.getPlayer();
+            ItemStack item = e.getItem();
+
+            if (!isItem(item))
```
Unnecessary check, since this inside an ItemHandler for the tool it is a given that this item matches. These checks are only necessary in Listener based handlers.

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,61 @@
+            if (!isItem(item))
+                return;
+
+            if(!e.getClickedBlock().isPresent()) return;    // Check if clicked block is null > return
```
Use open and close braces for all statements, I'd like it if my code was formatted following [Google's Style Guide](https://google.github.io/styleguide/javaguide.html) (This is also standard for Slimefun PRs so I just adopted their practices, also makes code look cleaner if everything is uniform)

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,61 @@
+            if(SlimefunPlugin.getProtectionManager().hasPermission(e.getPlayer(), b,
+                    ProtectableAction.BREAK_BLOCK)) // Check if player may 'break' the block to harvest
+
+            if(!(b.getBlockData() instanceof Ageable)) return;  // Check if clicked block is Ageable (and thus a crop), if not return
```
I believe crops such as Sugar Cane and Saplings will have problems with this, they are ageable but don't operate the same way as wheat does.

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,61 @@
+            if(crop.getAge() != crop.getMaximumAge()) return; // if crop is not fully grown return
+
+            // ready to harvest
+            for (ItemStack i : b.getDrops()) {
```
I'm not sure how to approach this yet, but it would be nice if effects such as fortune were applied on the drops as well.

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,61 @@
+            }
+            crop.setAge(CropState.SEEDED.ordinal());
+            b.setBlockData(crop, false);
+            item.setDurability((short) (item.getDurability()+1));
```
Since this is a Slimefun item, you can implement DamageableItem and use the corresponding methods to damage the item.

In regards to applying damage to an item, IIRC, setDurability is deprecated. You should be casting the ItemMeta to Damageable and then setting the damage from there.

Rene98 committed:
Changes:
- removed recipe output at item setup
- formatted lore of item
- changed lists to class constants
- made farmershoe not placeable
- reformatted if/else statements
- added support for sugar cane, cactus, chorus fruit and bamboo
- added support for enchantments to work with breaking crops
- added proper way to 'damage' item via slimefun

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,114 @@
+    }
+    public static final List<Material> breakOnly = Arrays.asList(Material.MELON, Material.PUMPKIN);
+    public static final List<Material> farmables = Arrays.asList(Material.DIRT, Material.GRASS_BLOCK, Material.COARSE_DIRT);
+    public static final List<Material> Spectallplants = Arrays.asList(Material.BAMBOO, Material.CHORUS_PLANT);
```
What is spectall? Also all constants should be FULL_CAPS.

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,114 @@
+        return true;
+    }
+
+    public Material getcrop(Material mat) {
```
Camelcase

NCBPFluffyBear replied:
Does this need to be public?

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,114 @@
+        super(category, item, recipeType, recipe);
+
+    }
+    public static final List<Material> breakOnly = Arrays.asList(Material.MELON, Material.PUMPKIN);
```
Constants should be listed above the constructor, do these need to be public?

NCBPFluffyBear commented:
Can you also upload a video of the tool working?

Rene98 commented:
https://www.youtube.com/watch?v=7vo7gD4fE-8

Rene98 committed:
- Code restructure
- re-done tall plant handling

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,108 @@
+
+public class FarmersHoe extends SimpleSlimefunItem<ItemUseHandler> implements NotPlaceable, DamageableItem {
+
+  static final List<Material> BREAKONLY = Arrays.asList(Material.MELON, Material.PUMPKIN);
```
Separate words with _ i.e. BREAK_ONLY and TALL_PLANTS

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,108 @@
+      Player p = e.getPlayer();
+      ItemStack item = e.getItem();
+
+
```
Remove extra line (Formatting)

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,108 @@
+        return;    // Check if clicked block is null > return
+      }
+      Block b = e.getClickedBlock().get();
+      if (FARMABLES.contains(b.getType())) {
```
Add a line break between each "action" that you do.

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/SupServItems.java:
```diff
@@ -72,4 +72,12 @@ private SupServItems() {}
+            Material.IRON_HOE,
+            "&3Farmers&8Hoe",
+            "",
+            "&7Makes sure harvested plots get reseeded, at the cost of durability",
```
Split this into two lines, the lore currently gets forced out of the screen for me
![image](https://user-images.githubusercontent.com/31554056/121469022-61943880-c981-11eb-9390-d610078a22e8.png)

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,108 @@
+        return;
+      }
+      e.cancel();
+      if (!SlimefunPlugin.getProtectionManager().hasPermission(e.getPlayer(), b,
```
According to Google's style guide, you can have lines up to 100 characters. I'm pretty sure this doesn't exceed that limit, so you can put it back in one line.

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,108 @@
+
+public class FarmersHoe extends SimpleSlimefunItem<ItemUseHandler> implements NotPlaceable, DamageableItem {
+
+  static final List<Material> BREAKONLY = Arrays.asList(Material.MELON, Material.PUMPKIN);
```
private

NCBPFluffyBear replied:
These could also probably be sets instead of lists

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,108 @@
+    };
+  }
+
+  @Override
```
Move this to the bottom of the class under all of your methods

NCBPFluffyBear commented on src/main/java/io/ncbpfluffybear/supserv/items/FarmersHoe.java:
```diff
@@ -0,0 +1,108 @@
+
+  public FarmersHoe(Category category, SlimefunItemStack item, RecipeType recipeType, ItemStack[] recipe) {
+    super(category, item, recipeType, recipe);
+
```
Remove extra line

NCBPFluffyBear reviewed (commented):
Some more stuff

NCBPFluffyBear commented:
Found an additional bug, when right clicking on the bottom of sugar cane it drops an extra sugar cane.
